src/main/java/org/zalando/nakadi/service/EventStreamWriterBinary.java

```diff
@@ -48,7 +68,7 @@ public int writeBatch(final OutputStream os, final Cursor cursor, final List<Str
         if (!events.isEmpty()) {
             os.write(B_EVENTS_ARRAY_BEGIN);
             for (int i = 0; i < events.size(); i++) {
-                final byte[] event = events.get(i).getBytes(StandardCharsets.UTF_8);
+                final byte[] event = extractor.apply(events.get(i)).getBytes(StandardCharsets.UTF_8);
```

Comment:
@adyach I removed casting by duplicating the code. It was not a "minimal" duplication but a full duplication.

As to Kafka byte array, this is not part of this PR. This PR is just about adapting another PR mentioned in the description. Further optimisations should be done in another ticket.